### ISSUE TITLE
Operator proposal: `defer`

### ISSUE THREAD

**Issue by velialiev** [effector/patronum#329]
# Motivation (the <s>fake</s> story behind the operator)

Web is asynchronous by it's nature and there are a lot of cases when we wait for something, defer/postpone something and so on.

Let's assume we have a user that want to pay using his wallet created in our app. He asks us how to do it and we want to give him an url that either will open wallet creation or wallet details depending on whether he already created the wallet or no.

Both of wallet creation and wallet details are dialogs belonging to the pay page. Existence of the wallet is checked by the request inited by pay page and buttons that opened this dialogs are disabled until data is loaded. Everything worked pretty well until we decided to give user a direct url that opens one of these dialogs. The problem is that the data needed to determine what dialog to open is still loading.

Sure we can just rework our UI to have one dialog that shows loading until data is loaded and renders different contents despite these dialogs are not related at all but the essence of the problem is that we have a synchronous event of opening our url and want to handle it asynchronously. In imperative API like effect handler we can do it with ease using async/await but there is no solution for declarative API

# Solution

Since almost everything is asynchronous in web, sometimes we want to handle events asynchronously. In imperative API like effect handler we can do it with ease using async/await, but there is no solution for declarative approach. We can rely on effect's done/fail events to call event after an asynchronous action but this works only in case when we have `event -> effect -> another event` chain. There is a case when event doesn't trigger effect but depends on data fetched by it. Effect in the other hand is called by another trigger.

In this case we have two branches of code:

1. If data **ALREADY** fetched then we should handle an event
2. If data **IS NOT** fetched **YET** then we should wait for data and then handle an event

That is where `defer` operator comes in.

```ts
sample({
  clock: defer({ clock: event, until: $itCanBeHandled }),
  target: handleEventFx,
})
```

You can read this as "when event is happend check if $itCanBeHandled is true. If so call the handleEventFx. If no wait until $itCanBeHandled became true and then call handleEventFx.

Let's try to solve the problem described in motivation. We want to open either wallet creation or wallet details dialog depending on whether wallet exists by direct url

```ts
split({
  clock: defer({ clock: event, until: equals(walletQuery.status, 'done') }),
  source: $wallet,
  match: (wallet) => (wallet ? 'hasWallet' : 'noWallet'),
  cases: {
    hasWallet: walletDialog.open,
    noWallet: walletCreationDialog.open,
  },
})
```

In this case we will run the split only when $wallet be ready to determine what dialog should be opened.

**Comment by velialiev** [comment]
I've added implementation draft: https://github.com/effector/patronum/pull/330

If proposal would be approved I'll add tests and docs there

**Comment by victordidenko** [comment]
In our codebase we use this simple operator:
```ts
export function postpone<T>({
  clock,
  until,
  target,
}: {
  clock: Event<T>
  until: Store<boolean>
  target: Event<T>
}): Event<T> {
  return sample({
    clock: [clock, until],
    source: clock,
    filter: until,
    target,
  })
}
```
Looks like you are proposing the same?

**Comment by velialiev** [comment]
> In our codebase we use this simple operator:
> 
> ```ts
> export function postpone<T>({
>   clock,
>   until,
>   target,
> }: {
>   clock: Event<T>
>   until: Store<boolean>
>   target: Event<T>
> }): Event<T> {
>   return sample({
>     clock: [clock, until],
>     source: clock,
>     filter: until,
>     target,
>   })
> }
> ```
> 
> Looks like you are proposing the same?

If i got it right your operator will call the `target` when `until` is changed from false to true without calling clock at all. The operator i propose will call target ONLY when clock is called

**Comment by victordidenko** [comment]
> even if clock is not called yet

No, `sample` will not trigger target if event in source was not triggered

<img width="835" alt="image" src="https://github.com/effector/patronum/assets/15016227/c10725b9-93fc-4aa7-ac59-76d037e611a2">

But this operator will not reset its state, as your variant

**Comment by velialiev** [comment]
> > even if clock is not called yet
> 
> No, `sample` will not trigger target if event in source was not triggered
> 
> <img alt="image" width="835" src="https://private-user-images.githubusercontent.com/15016227/331196434-c10725b9-93fc-4aa7-ac59-76d037e611a2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTU4NjMxMDksIm5iZiI6MTcxNTg2MjgwOSwicGF0aCI6Ii8xNTAxNjIyNy8zMzExOTY0MzQtYzEwNzI1YjktOTNmYy00YWE3LWFjNTktNzZkMDM3ZTYxMWEyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTE2VDEyMzMyOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWY0MTFlN2Y5N2FlMzg4YWM2YzUwN2JkM2RiNWQ4YzE0NGIxOTJhZGRiZDYyNzU1Y2NhNTg0YjE5MmQ1MDA0ZTQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.OtwOj1z-gEBpxl61Fd5c-7_aNhK8TC1mWBKafgquj2A">
> But this operator will not reset its state, as your variant

Wow! Great catch :) It's first time I see someone using this limitation as the feature. Actually I see using event as the source second time including your case

But yeah, as you've mentioned above your solution will not reset its state and will break starting from second call since `until` will trigger the sample because source has already been called once

**Comment by illright** [comment]
This operator would be useful to me. Here's the use case I have:

I have a store that often receives updates, every new update basically invalidates the previous states. I want to process these updates using an asynchronous effect, but to avoid calling the processing effect too many times, I would like to defer processing a new update until the processing of the old update is finished. It's almost the same logic as `filter` in `sample`, except I need the update not to get lost, but rather the processing of it to be deferred.

This cannot be achieved using the method proposed above because my store is not a one-off, so the event will be called repeatedly.

**Comment by illright** [comment]
Copy-pasted the operator from the PR, works great. Here's the solution that worked for me:
```ts
const treeChanged = debounce(vfs.$tree, 500)
const runRulesFx = createEffect(runRules)

sample({
  clock: defer({ clock: treeChanged, until: not(runRulesFx.pending) }),
  target: runRulesFx,
})
```
